interface_tester/interface_test.py

```diff
@@ -379,22 +380,21 @@ def _run(self, event: Union[str, Event]):
         # some required config, a "happy" status, network information, OTHER relations.
         # Typically, should NOT touch the relation that this interface test is about
         #  -> so we overwrite and warn on conflict: state_template is the baseline,
-        state = (self.ctx.state_template or State()).copy()
+        state = copy.deepcopy(self.ctx.state_template or State())
 
         relations = self._generate_relations_state(
             state, input_state, self.ctx.supported_endpoints, self.ctx.role
         )
         # State is frozen; replace
-        modified_state = state.replace(relations=relations)
+        modified_state = dataclasses.replace(state, relations=relations)
 
         # the Relation instance this test is about:
         relation = next(filter(lambda r: r.interface == self.ctx.interface_name, relations))
-        evt: Event = self._cast_event(event, relation)
 
-        logger.info("collected test for %s with %s" % (self.ctx.interface_name, evt.name))
-        return self._run_scenario(evt, modified_state)
+        logger.info("collected test for %s with %s" % (self.ctx.interface_name, event))
```

Comment:
why this change?

Reply:
I moved the `_cast_event` call into `_run_scenario` (I don't remember exactly why - I could figure that out if you like, or try to avoid it). Because of that, at this point `event` is `str|_Event` so the `.name` may not work.

I could also just move the logging line to be in `_run_scenario` also.

Reply:
uh. Well in general I try to coerce types as soon as possible to keep the private signatures nice and simple, so if it's not too much trouble...

Reply:
Sure. It was to get access to `ctx.on`, but I've changed it to use `CharmEvents` directly instead, so it's much more minimal.